File: scripts/upload-to-turso.py
#!/usr/bin/env python3
"""
Upload local saints_encyclopedia.db to a Turso database.

Usage:
    Set env vars TURSO_DATABASE_URL and TURSO_AUTH_TOKEN, then run:
    python scripts/upload-to-turso.py

    Or pass them as arguments:
    python scripts/upload-to-turso.py --url libsql://... --token ...
"""

import sqlite3
import argparse
import os
import sys
import json
from pathlib import Path
from urllib.request import urlopen, Request
from urllib.error import HTTPError

DB_PATH = Path(__file__).parent.parent / "scraper" / "saints_encyclopedia.db"


def turso_execute(url: str, token: str, statements: list[dict]) -> dict:
    """Execute statements via Turso HTTP API."""
    # Convert libsql:// to https://
    http_url = url.replace("libsql://", "https://")
    if not http_url.startswith("https://"):
        http_url = f"https://{http_url}"
    endpoint = f"{http_url}/v3/pipeline"

    body = json.dumps({
        "requests": [
            {"type": "execute", "stmt": {"sql": s["sql"]}} for s in statements
        ] + [{"type": "close"}]
    }).encode()

    req = Request(
        endpoint,
        data=body,
        headers={
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json",
        },
    )

    try:
        with urlopen(req) as resp:
            return json.loads(resp.read())
    except HTTPError as e:
        error_body = e.read().decode() if e.fp else ""
        raise RuntimeError(f"HTTP {e.code}: {error_body[:500]}") from e


def main():
    parser = argparse.ArgumentParser(description="Upload SQLite DB to Turso")
    parser.add_argument("--url", default=os.environ.get("TURSO_DATABASE_URL", ""))
    parser.add_argument("--token", default=os.environ.get("TURSO_AUTH_TOKEN", ""))
    parser.add_argument("--db", default=str(DB_PATH))
    args = parser.parse_args()

    if not args.url or not args.token:
        print("Error: Set TURSO_DATABASE_URL and TURSO_AUTH_TOKEN env vars")
        print("  or pass --url and --token arguments")
        sys.exit(1)

    db_path = Path(args.db)
    if not db_path.exists():
        print(f"Error: Database not found at {db_path}")
        sys.exit(1)

    print(f"Source DB: {db_path}")
    print(f"Target:    {args.url}")

    # Dump the database
    conn = sqlite3.connect(str(db_path))
    statements = []
    for line in conn.iterdump():
        if line.startswith("BEGIN") or line.startswith("COMMIT"):
            continue
        statements.append({"sql": line})
    conn.close()

    print(f"Total statements: {len(statements)}")

    # Execute in batches
    BATCH_SIZE = 50
    executed = 0
    errors = 0

    for i in range(0, len(statements), BATCH_SIZE):
        batch = statements[i : i + BATCH_SIZE]
        try:
            turso_execute(args.url, args.token, batch)
            executed += len(batch)
        except Exception as e:
            # Try one by one on failure
            for stmt in batch:
                try:
                    turso_execute(args.url, args.token, [stmt])
                    executed += 1
                except Exception as inner_e:
                    msg = str(inner_e)
                    if "already exists" not in msg:
                        errors += 1
                        if errors <= 5:
                            print(f"\n  Error: {msg[:150]}")
                    executed += 1

        pct = (executed / len(statements)) * 100
        print(f"\r  Progress: {executed}/{len(statements)} ({pct:.0f}%) - {errors} errors", end="", flush=True)

    print(f"\n\nDone! {executed} statements, {errors} errors.")

    # Verify
    try:
        result = turso_execute(args.url, args.token, [{"sql": "SELECT COUNT(*) as cnt FROM games"}])
        results = result.get("results", [])
        if results:
            rows = results[0].get("response", {}).get("result", {}).get("rows", [])
            if rows:
                print(f"Verification: {rows[0][0].get('value', '?')} games in Turso")
    except Exception as e:
        print(f"Verification query failed: {e}")


if __name__ == "__main__":
    main()

File: web/scripts/upload-to-turso.mjs
#!/usr/bin/env node
/**
 * Upload local saints_encyclopedia.db to Turso.
 *
 * Usage:
 *   node scripts/upload-to-turso.mjs
 *
 * Requires env vars: TURSO_DATABASE_URL, TURSO_AUTH_TOKEN
 * Or pass them as arguments:
 *   TURSO_DATABASE_URL=libsql://... TURSO_AUTH_TOKEN=... node scripts/upload-to-turso.mjs
 */

import { createClient } from "@libsql/client/web";
import { readFileSync } from "fs";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";
import { execSync } from "child_process";

const __dirname = dirname(fileURLToPath(import.meta.url));

const TURSO_URL = process.env.TURSO_DATABASE_URL;
const TURSO_TOKEN = process.env.TURSO_AUTH_TOKEN;

if (!TURSO_URL || !TURSO_TOKEN) {
  console.error("Error: Set TURSO_DATABASE_URL and TURSO_AUTH_TOKEN env vars");
  process.exit(1);
}

const DB_PATH = resolve(__dirname, "../../scraper/saints_encyclopedia.db");

console.log(`Source DB: ${DB_PATH}`);
console.log(`Target: ${TURSO_URL}`);

// Use sqlite3 CLI to dump the database as SQL
console.log("\nDumping local database to SQL...");
let sqlDump;
try {
  sqlDump = execSync(`sqlite3 "${DB_PATH}" .dump`, {
    maxBuffer: 100 * 1024 * 1024,
    encoding: "utf-8",
  });
} catch (e) {
  console.error(
    "Error: sqlite3 CLI not found. Install SQLite3 or use an alternative method."
  );
  console.error("On Windows: winget install SQLite.SQLite");
  console.error(
    "\nAlternative: Use the Turso dashboard to upload the .db file directly."
  );
  process.exit(1);
}

console.log(`SQL dump size: ${(sqlDump.length / 1024 / 1024).toFixed(1)} MB`);

// Connect to Turso
const db = createClient({
  url: TURSO_URL,
  authToken: TURSO_TOKEN,
});

// Split into individual statements
const statements = sqlDump
  .split(";\n")
  .map((s) => s.trim())
  .filter((s) => s.length > 0 && !s.startsWith("--"))
  .map((s) => (s.endsWith(";") ? s : s + ";"));

console.log(`\nExecuting ${statements.length} statements...`);

// Execute in batches
const BATCH_SIZE = 20;
let executed = 0;
let errors = 0;

for (let i = 0; i < statements.length; i += BATCH_SIZE) {
  const batch = statements.slice(i, i + BATCH_SIZE);
  try {
    await db.batch(batch.map((sql) => ({ sql, args: [] })));
    executed += batch.length;
  } catch (e) {
    // Try statements individually on batch failure
    for (const sql of batch) {
      try {
        await db.execute({ sql, args: [] });
        executed++;
      } catch (innerErr) {
        // Skip expected errors like "table already exists"
        const msg = String(innerErr);
        if (!msg.includes("already exists") && !msg.includes("TRANSACTION")) {
          errors++;
          if (errors <= 5) {
            console.error(
              `  Error: ${msg.substring(0, 100)}...`
            );
          }
        }
        executed++;
      }
    }
  }

  if (executed % 200 === 0 || i + BATCH_SIZE >= statements.length) {
    process.stdout.write(
      `\r  Progress: ${executed}/${statements.length} statements (${errors} errors)`
    );
  }
}

console.log(`\n\nDone! Executed ${executed} statements with ${errors} errors.`);

// Verify
const result = await db.execute("SELECT COUNT(*) as count FROM games");
console.log(`\nVerification: ${result.rows[0].count} games in Turso database`);

const players = await db.execute("SELECT COUNT(*) as count FROM players");
console.log(`             ${players.rows[0].count} players in Turso database`);

process.exit(0);
